### PR TITLE
Add function toDaysSinceYearZero()

### LINKED ISSUE #54277: MySQL compatibility: TO_DAYS function
This one is a minor one but is still used by Tableau online when converting DateTime columns during the table column types introspection stage.

image

Sample generated query:
```sql
SELECT `cell_towers`.`area`                        AS `area`,
       `cell_towers`.`averageSignal`               AS `averageSignal`,
       `cell_towers`.`cell`                        AS `cell`,
       `cell_towers`.`changeable`                  AS `changeable`,
       `cell_towers`.`created`                     AS `created`,
       `cell_towers`.`lat`                         AS `lat`,
       `cell_towers`.`lon`                         AS `lon`,
       `cell_towers`.`mcc`                         AS `mcc`,
       `cell_towers`.`net`                         AS `net`,
       `cell_towers`.`radio`                       AS `radio`,
       `cell_towers`.`range`                       AS `range`,
       `cell_towers`.`samples`                     AS `samples`,
       `cell_towers`.`unit`                        AS `unit`,
       (TO_DAYS(`cell_towers`.`updated`) - 693961) AS `updated`
FROM `cell_towers`
LIMIT 100
```

which fails, because

```
Code: 46. DB::Exception: Unknown function TO_DAYS
```

**How to reproduce**

* Which ClickHouse server version to use: latest master
* Which interface to use, if matters: MySQL
* Sample data for all these tables: cell_towers sample dataset

**Expected behavior**

TO_DAYS function works like in MySQL

### PR DESCRIPTION
**Pull request by rschu1ze**
Fixes: #54277

Review note: https://github.com/ClickHouse/ClickHouse/pull/54479/commits/0a05831839d552488cf77b4c6d6eb00a74308f21 contains unrelated cleanups

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added function "toDaysSinceYearZero" with alias "TO_DAYS()" (for compatibility with MySQL) which returns the number of days passed since 0001-01-01.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit a384e0c46a355c2be2b51f2f1ddac6545970be1e with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54479/a384e0c46a355c2be2b51f2f1ddac6545970be1e/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by alexey-milovidov** [comment]
It's strange that the function is named ...SinceYearZero, but you say it counts from the year 0001.
Note: if I recall correctly, year zero did not exist... and the year 1 did not exist as well, it was just backfilled in the calendar later in a way that after year -1 there is year 1, skipping 0. Nevertheless, we should resolve this confusion by naming the function better. For example, `toDaysSinceFirstYear`

File: src/Functions/DateTimeTransforms.h [hunk 1]
```diff
@@ -672,32 +672,32 @@ struct ToStartOfHourImpl
...
+    static UInt32 execute(UInt32 t, const DateLUTImpl & time_zone)
     {
         return time_zone.toStartOfHour(t);
     }
 
-    static inline UInt32 execute(Int32, const DateLUTImpl &)
+    static UInt32 execute(Int32, const DateLUTImpl &)
     {
-        throwDateTimeIsNotSupported(name);
+        throwDateIsNotSupported(name);
```
Comment (hanfei1991) [inline src/Functions/DateTimeTransforms.h:687]: Seems Int32 / Uint64 stands for date, and UInt32/ Int64 ..stands for datetime?

Can we use template to connect field type and `date/datetime` in order not to confuse them?

Comment (rschu1ze) [reply src/Functions/DateTimeTransforms.h:687]: I think so, but that would be a separate PR.

**Comment by rschu1ze** [comment]
@alexey-milovidov True, year 0 doesn't exist in common understanding, it only exists in the astronomical ("scientific") calendar where 1 AC = 1, 1 BC = 0, 2 BC = -1 etc. Contrary to common understanding, MySQL's `TO_DAYS` uses astronomical year 0 as reference point, and they use the proleptic calendar for simplicity, i.e. ignore the 10 days "lost" in October 1582. I'll 1. document it and 2. change the offset constant for 1900-01-01 which I originally took from some online calculator, now using the MySQL offset for 100% compatibility.

**Review by hanfei1991 — APPROVED** [review]

**Pull request merged**